src/agent.rs

```diff
@@ -88,10 +88,13 @@ pub trait Session: 'static + Sync + Send + Sized {
     {
         loop {
             if let Some(incoming_message) = adapter.try_next().await? {
-                let response = self.handle(incoming_message).await.map_err(|e| {
-                    error!("Error handling message; error = {:?}", e);
-                    AgentError::User
-                })?;
+                let response = match self.handle(incoming_message).await {
+                    Ok(message) => message,
+                    Err(e) => {
+                        error!("Error handling message: {:?}", e);
+                        Message::Failure
```

Comment:
it kind of depends on the error I think.
It should return a `Failure` on a `AgentError::User` but I would close on a `AgentError::Proto` or `AgentError::Io`

Reply:
Good point :+1: I'll add the appropriate handling code.